File: backend/documentservice/src/main/java/com/example/documentService/parser/factory/ParserFactory.java
package com.example.documentService.parser.factory;

import com.example.documentService.parser.PDFParser;
import com.example.documentService.parser.TextParser;
import com.example.documentService.parser.WordParser;
import com.example.documentService.parser.interfaces.DocumentParser;
import org.springframework.stereotype.Component;

@Component
public class ParserFactory {
    private final PDFParser pdfParser;
    private final WordParser wordParser;
    private final TextParser textParser;

    public ParserFactory(PDFParser pdfParser, WordParser wordParser, TextParser textParser){
        this.pdfParser = pdfParser;
        this.wordParser = wordParser;
        this.textParser = textParser;
    }
    public DocumentParser getParser(String fileType){
        if (fileType == null || fileType.isEmpty()){
            throw new IllegalArgumentException("File type cannot be null or empty");
        }
        switch (fileType.toLowerCase()){
            case "pdf":
                return new PDFParser();
            case "doc":
                return new WordParser();
            case "txt":
                return new TextParser();
            default:
                throw new IllegalArgumentException("Unsupported file type: " + fileType);
        }
    }
}

File: ai-service/Document_Parser/Parsers/parser_factory.py
from .base_parser import DocumentParser
from .pdf_parser import PDFParser
from .word_parser import WordParser
from .text_parser import TextParser


class DocumentParserFactory:
    @staticmethod
    def get_parser(file_type: str) -> DocumentParser:
        if file_type == "pdf":
            return PDFParser()
        elif file_type in ['docx', 'doc']:
            return WordParser()
        elif file_type in ['text', 'txt']:
            return TextParser()
        else:
            raise ValueError(f"Unsupported file type: {file_type}")
